server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java

```diff
@@ -0,0 +1,185 @@
+  public static final String PARAM_DESCRIPTION = "description";
+  public static final String PARAM_DOMAIN = "domain";
+
+  private static final String FIELD_ID = "id";
+  private static final String FIELD_NAME = PARAM_NAME;
+  private static final String FIELD_KEY = PARAM_KEY;
+  private static final String FIELD_TYPE = PARAM_TYPE;
+  private static final String FIELD_DESCRIPTION = PARAM_DESCRIPTION;
+  private static final String FIELD_DOMAIN = PARAM_DOMAIN;
+
+  private final DbClient dbClient;
+  private final UserSession userSession;
+
+  public CreateAction(DbClient dbClient, UserSession userSession) {
+    this.dbClient = dbClient;
+    this.userSession = userSession;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction(ACTION)
+      .setPost(true)
+      .setDescription("Create custom metric.<br /> Requires 'Administer System' permission.")
+      .setSince("5.2")
+      .setHandler(this);
+
+    action.createParam(PARAM_NAME)
+      .setRequired(true)
+      .setDescription("Name")
+      .setExampleValue("Team Size");
+
+    action.createParam(PARAM_KEY)
+      .setRequired(true)
+      .setDescription("Key")
+      .setExampleValue("team_size");
+
+    action.createParam(PARAM_TYPE)
+      .setRequired(true)
+      .setDescription("Metric type key")
+      .setPossibleValues(Metric.ValueType.names())
+      .setExampleValue(Metric.ValueType.INT.name());
+
+    action.createParam(PARAM_DESCRIPTION)
+      .setDescription("Description")
+      .setExampleValue("Size of the team");
+
+    action.createParam(PARAM_DOMAIN)
+      .setDescription("Domain")
+      .setExampleValue("Tests");
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    userSession.checkLoggedIn().checkGlobalPermission(GlobalPermissions.SYSTEM_ADMIN);
+    String key = request.mandatoryParam(PARAM_KEY);
+
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      MetricDto metricInDb = dbClient.metricDao().selectNullableByKey(dbSession, key);
+      checkMetricInDb(metricInDb, key);
+
+      MetricDto metricTemplate = newMetricTemplate(request);
+
+      if (metricInDb == null) {
+        metricInDb = insertNewMetric(dbSession, metricTemplate);
+      } else {
+        updateMetric(dbSession, metricInDb, metricTemplate);
+      }
+
+      JsonWriter json = response.newJsonWriter();
+      writeMetric(json, metricInDb);
+      json.close();
+    } finally {
+      MyBatis.closeQuietly(dbSession);
+    }
+  }
+
+  private MetricDto newMetricTemplate(Request request) {
+    MetricDto metricTemplate = new MetricDto()
+      .setKey(request.mandatoryParam(PARAM_KEY))
+      .setShortName(request.mandatoryParam(PARAM_NAME))
+      .setValueType(Metric.ValueType.valueOf(request.mandatoryParam(PARAM_TYPE)).name());
+    if ((request.param(PARAM_DOMAIN) != null)) {
+      metricTemplate.setDomain(request.param(PARAM_DOMAIN));
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Possible null pointer dereference in org.sonar.server.metric.ws.CreateAction.newMetricTemplate(Request) due to return value of called method [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE)